ops/pebble.py

```diff
@@ -2584,6 +2587,7 @@ def exec(
             APIError: if an error occurred communicating with pebble, or if the command is not
                 found.
             ExecError: if the command exits with a non-zero exit code.
+            TimeoutError: if the maximum timeout is reached.
```

Comment:
I think `ExecProcess`'s `wait` and `wait_output` could also raise `TimeoutError`, and I think also the various action services methods of the Pebble client?

Reply:
`wait`/`wait_output` did call `wait_change` which calls `_wait_change_using_wait`, but `wait_change` does not handle timeout exception, so `wait`/`wait_output` raise a `ChangeError` instead (already in their docstring), see my other reply in the other comment.

Reply:
Ah, I remember now that @benhoyt and I discussed all this already last year in #908. We decided that it's too noisy to say that basically all the Pebble/Container methods might raise ChangeError/Timeout error, so:

* In [Container](https://ops.readthedocs.io/en/latest/#ops.Container) we say:

> For methods that make changes to the container, if the change fails or times out, then a [ops.pebble.ChangeError](https://ops.readthedocs.io/en/latest/#ops.pebble.ChangeError) or [ops.pebble.TimeoutError](https://ops.readthedocs.io/en/latest/#ops.pebble.TimeoutError) will be raised.

* In [pebble.Client](https://ops.readthedocs.io/en/latest/#ops.pebble.Client) we say:

> For methods that wait for changes, such as [start_services()](https://ops.readthedocs.io/en/latest/#ops.pebble.Client.start_services) and [replan_services()](https://ops.readthedocs.io/en/latest/#ops.pebble.Client.replan_services), if the change fails or times out, then a [ChangeError](https://ops.readthedocs.io/en/latest/#ops.pebble.ChangeError) or [TimeoutError](https://ops.readthedocs.io/en/latest/#ops.pebble.TimeoutError) will be raised.

So perhaps we ought to be rejecting #1131 instead (otherwise we are going to go down the route of adding them all again). We *do* say what happens when `exec` times out, you just have to read the `Container` doc to see it.

Reply:
Ah yes, I think Tony's right -- thanks!